### PR TITLE
using encoder instead of texture operation to avoid access issues.

### PR DESCRIPTION
**Pull request by hana-alice**
> replaceRegion:mipmapLevel:slice:withBytes:bytesPerRow:bytesPerImage:

> This method immediately copies the pixel data into the texture. It doesn't synchronize against any GPU accesses to the texture. For example, if a command buffer includes read or write operations on a given texture, you must ensure that these operations complete before calling the replaceRegion:mipmapLevel:slice:withBytes:bytesPerRow:bytesPerImage: method on the given texture.

### CODE REVIEW

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm [hunk 1]
```diff
@@ -365,12 +365,23 @@
...
+        
+        [encoder copyFromBuffer:stagingBuffer.mtlBuffer
+                   sourceOffset:stagingBuffer.startOffset
+              sourceBytesPerRow:bytesPerRow
+            sourceBytesPerImage:bytesPerImage
+                     sourceSize:stagingRegion.sourceSize
+                      toTexture:dstTexture
+               destinationSlice:stagingRegion.destinationSlice
+               destinationLevel:stagingRegion.destinationLevel
+              destinationOrigin:stagingRegion.destinationOrigin];
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLCommandBuffer.mm:384]: 在 iOS 上跑一下纹理压缩的测试例子，看是否有问题。

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm [hunk 2]
```diff
@@ -365,12 +365,23 @@
...
-                            slice:stagingRegion.destinationSlice
-                        withBytes:convertedData
-                      bytesPerRow:bytesPerRow
-                    bytesPerImage:bytesPerImage];
+        
+        CCMTLGPUBuffer stagingBuffer;
+        uint channelPerPixel = cc::gfx::GFX_FORMAT_INFOS[(uint)format].size;
+        //size = w * h * sizeof(type of data stored in) * channelNum;
+        stagingBuffer.size = bufferSize[i] * sizeof(std::remove_pointer<decltype(convertedData)>::type) * channelPerPixel;
+        _mtlDevice->gpuStagingBufferPool()->alloc(&stagingBuffer);
```
Comment (Mee-gu) [inline cocos/renderer/gfx-metal/MTLCommandBuffer.mm:373]: 需要考虑 alignment 对齐

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm [hunk 3]
```diff
@@ -365,12 +365,23 @@
...
-                      mipmapLevel:stagingRegion.destinationLevel
-                            slice:stagingRegion.destinationSlice
-                        withBytes:convertedData
-                      bytesPerRow:bytesPerRow
-                    bytesPerImage:bytesPerImage];
+        
+        CCMTLGPUBuffer stagingBuffer;
+        uint channelPerPixel = cc::gfx::GFX_FORMAT_INFOS[(uint)format].size;
+        //size = w * h * sizeof(type of data stored in) * channelNum;
+        stagingBuffer.size = bufferSize[i] * sizeof(std::remove_pointer<decltype(convertedData)>::type) * channelPerPixel;
```
Comment (Mee-gu) [inline cocos/renderer/gfx-metal/MTLCommandBuffer.mm:372]: gfx::FormatSize 封装了这个过程，不用自己算

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm [hunk 4]
```diff
@@ -365,12 +365,23 @@
...
+        _mtlDevice->gpuStagingBufferPool()->alloc(&stagingBuffer);
+        memcpy(stagingBuffer.mappedData, convertedData, stagingBuffer.size);
```
Comment (Mee-gu) [inline cocos/renderer/gfx-metal/MTLCommandBuffer.mm:374]: 需要考虑 offset，不然每次都是更新同一个位置

Comment (hana-alice) [reply cocos/renderer/gfx-metal/MTLCommandBuffer.mm:374]: stagingBuffer.mappedData要更新的数据内容，经由[encoder copyFromBuffer : sourceSize : destinationOrigin]定位更新的位置。
 
convertedData来自于：

> void CCMTLCommandBuffer::copyBuffersToTexture(const uint8_t *const *buffers, Texture *texture, const BufferTextureCopy *regions, uint count)

const auto convertedData = mu::convertData(buffers[i], bufferSize[i], format);
buffers[i]这里是已经有过偏移的

Comment (Mee-gu) [reply cocos/renderer/gfx-metal/MTLCommandBuffer.mm:374]: 恩，我看错了

**Comment by hana-alice** [comment]
blitCommandBuffer要更新PVR格式的内容，要求原数据是linear排列的[1]，但是有些PVR并非如此(Morton)[2]，加载的时候不匹配会直接崩掉。
[1]https://developer.apple.com/documentation/metal/mtlblitoption/mtlblitoptionrowlinearpvrtc?language=occ
[2]http://cdn.imgtec.com/sdk-documentation/PVRTC%20Specification%20and%20User%20Guide.pdf

**Pull request closed without merging**